File: aws-part/cdk/part2_stack.py
from typing import Optional

import aws_cdk as cdk
from aws_cdk import (
    Stack,
    aws_s3 as s3,
    aws_dynamodb as dynamodb,
)
from constructs import Construct


class Part2Stack(Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        existing_table_name: Optional[str] = None,
        image_bucket_name: Optional[str] = None,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # 最小: 空のS3バケット（削除容易な開発用設定）
        self.bucket = s3.Bucket(
            self,
            "ImageBucket",
            bucket_name=image_bucket_name,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            auto_delete_objects=True,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            enforce_ssl=True,
        )

        # 既存DynamoDBを参照（存在する場合のみ）
        table_name = existing_table_name or self.node.try_get_context("ddbTableName")
        if table_name:
            self.existing_table = dynamodb.Table.from_table_name(
                self, "ExistingDynamoTable", table_name=table_name
            )

        # ここに後で: S3->Lambda->Bedrock の配線を追加予定

File: aws-part/cdk/lib/part2-stack.js
import * as cdk from 'aws-cdk-lib';
import { 
  aws_s3 as s3,
  aws_dynamodb as dynamodb,
  aws_lambda as lambda,
  aws_lambda_nodejs as lambdaNodejs,
  aws_iam as iam,
} from 'aws-cdk-lib';
import { aws_s3_notifications as s3n } from 'aws-cdk-lib';
import path from 'node:path';

export class Part2Stack extends cdk.Stack {
  constructor(scope, id, props = {}) {
    super(scope, id, props);

    const { imageBucketName, ddbTableName } = props;

    if (!imageBucketName) {
      throw new Error('IMAGE_BUCKET_NAME is required (existing S3 bucket)');
    }
    if (!ddbTableName) {
      throw new Error('DDB_TABLE_NAME is required (existing DynamoDB table)');
    }

    // 先に作ったS3とDynamoDBの情報を取得します。
    const bucket = s3.Bucket.fromBucketName(this, 'ExistingImageBucket', imageBucketName);
    const table = dynamodb.Table.fromTableName(this, 'ExistingDynamoTable', ddbTableName);

    // 画像の説明を生成するLambda関数を作成します。
    const describeFn = new lambdaNodejs.NodejsFunction(this, 'ImageDescribeFn', {
      entry: path.resolve('lambda/part2-handler.js'),
      runtime: lambda.Runtime.NODEJS_LATEST,
      memorySize: 512,
      timeout: cdk.Duration.seconds(60),
      environment: {
        TABLE_NAME: ddbTableName,
        DDB_PK: 'image_id',
      },
    });

    new lambda.EventInvokeConfig(this, 'ImageDescribeFnInvokeConfig', {
      function: describeFn,
      retryAttempts: 0,
    });

    bucket.grantRead(describeFn);
    table.grantReadWriteData(describeFn);
    describeFn.addToRolePolicy(new iam.PolicyStatement({ actions: ['bedrock:InvokeModel'], resources: ['*'] }));

    // 画像がS3にアップロードされたら、Lambdaに通知して、関数の起動を行う設定を追加します。
    const dest = new s3n.LambdaDestination(describeFn);
    bucket.addEventNotification(s3.EventType.OBJECT_CREATED_PUT, dest, { prefix: 'uploads/' });
  }
}
